django/db/migrations/recorder.py

```diff
@@ -47,16 +47,21 @@ def __str__(self):
 
     def __init__(self, connection):
         self.connection = connection
+        self._has_table = False
 
     @property
     def migration_qs(self):
         return self.Migration.objects.using(self.connection.alias)
 
     def has_table(self):
         """Return True if the django_migrations table exists."""
-        with self.connection.cursor() as cursor:
-            tables = self.connection.introspection.table_names(cursor)
-        return self.Migration._meta.db_table in tables
+        if not self._has_table:
```

Comment:
Do you think is there any way to ensure this improvement is not regressed by adding a test?

Reply:
Could we possibly add a test that uses `assertNumQueries` to ensure that multiple accesses are cached?

Reply:
For example:
```diff
diff --git a/tests/migrations/test_loader.py b/tests/migrations/test_loader.py
index 800d715bc9..d1b63b5e9a 100644
--- a/tests/migrations/test_loader.py
+++ b/tests/migrations/test_loader.py
@@ -48,6 +48,15 @@ class RecorderTests(TestCase):
             set(),
         )
 
+    def test_has_table_cached(self):
+        recorder = MigrationRecorder(connection)
+        with self.assertNumQueries(2):
+            recorder.record_applied("myapp", "0432_ponies")
+        # has_table() doesn't perform queries when the existense of the
+        # django_migrations table is confirmed.
+        with self.assertNumQueries(1):
+            recorder.record_applied("myapp", "0432_ponies")
+
 
 class LoaderTests(TestCase):
     """

```

Reply:
Thanks! I'll add some tests and a comment now

Reply:
> Thanks! I'll add some tests and a comment now

See https://github.com/django/django/pull/16892#discussion_r1205317875

Reply:
I've added tests for the `has_table()` method instead of `record_applied()` as I feel like the number of queries is less likely to change with any future work.

I had to explicitly delete the `Migration` object at the start of the test, I'm not sure if there is a better way to do this.